geps/gep-1958.md

```diff
@@ -0,0 +1,44 @@
+# GEP-1958: TLS ALPN Routing
+Issue: #1958
+Status: Provisional
+## TLDR
+Add support for routing traffic based on the application protcols present in the TLS client hello of a TLS connection.
+
+## Goals
```

Comment:
One thing we may need to take into account is not just routing based on the ALPN, but what ALPNs we should offer in the first place.

It doesn't help to have a ALPN route for `foobar` if the Gateway doesn't advertise it

Reply:
Good point, I think that depends on whether TLS termination or TLS passthrough is being used.

I was only considering the passthrough case where the gateway is not responsible for the response. In this case the gateway is only making routing decisions, its up to the service owner to ensure the given protocol is actually selected by the upstream service.

For the TLS termination case I suppose the gateway would need to be configured to route and select ALPN based on the values configured by the service owner.

Reply:
Ah right, forgot that. I think TLS route is only for passthrough today (?) so that would solve that concern

Reply:
I suspect we need ALPN both for TLS passthrough and for termination, and likely the APIs will be quite different. 

For passthrough - would be good to have some example use cases, I have not seen any server doing this and I'm not sure what will the user do - for example if alpn is 'h2', would the stream be routed to a different backend ? Doesn't seem a common enough use case.

 For 'terminate' - beyond an ordered list of protocols that the server accepts, not sure what else we can do. Server doesn't advertise protocols - it needs to selects one from the list sent by the client.

Besides h2 - I have not seen a lot of broadly used ALPNs on the internet.

Reply:
I think the two use cases and APIs are different enough that it makes sense to limit the scope this GEP to only include the tls passthrough use case.

Reply:
Given the lack of response, we'll consider this comment resolved for the moment. Please mark it unresolved if more conversation is needed.